### ISSUE TITLE
Add Firestore indexes to local emulator

### ISSUE THREAD

**Issue by nicholaschiang** [firebase/firebase-tools#2027]
### [REQUIRED] Environment info

**firebase-tools:**
7.15.0

**Platform:**
Ubuntu 18.04.4
### [REQUIRED] Test case

I'm trying to test [our Firestore indexes](https://tutorbook.app/docs/tutorial-Storage.html), but they seem to always be working (regardless of the indexing settings included in `firestore.indexes.json`). Below is the [test case in question](https://github.com/tutorbookapp/tutorbook/blob/develop/firebase/test/test.js#L382):
```javascript
const FILTERS = {
    'grade': ['==', TUTOR.grade],
    'gender': ['==', TUTOR.gender],
    'location': ['==', TUTOR.location],
    'type': ['==', TUTOR.type],
    'payments.type': ['==', TUTOR.payments.type],
    'config.showProfile': ['==', true],
};
const SORTERS = [
    'avgRating',
    'numRatings',
];

describe('Tutorbook\'s Database Indexing', () => {
    const db = authedApp({
        uid: PUPIL.uid,
        email: PUPIL.email,
        access: PUPIL.access,
    });
    const filterCombos = combinations(Object.keys(FILTERS));
    return Promise.all(filterCombos.map(filters => it('lets users filter ' +
        'profiles by ' + filters.join(', '), () => {
            var query = db.collection('users')
                .where('access', 'array-contains-any', PUPIL.access);
            filters.map(filter => query =
                query.where(filter, FILTERS[filter][0], FILTERS[filter][1]));
            return Promise.all(SORTERS.map(sorter =>
                firebase.assertSucceeds(query.orderBy(sorter).get())));
        })));
});
```

### [REQUIRED] Steps to reproduce

Just try testing any type of composite or collection group query that would (normally) require an index. It will work on the emulator suite regardless of the existence of the required index (but it won't work in production without the index).

### [REQUIRED] Expected behavior

These tests should fail unless there are Firestore indexes (specified in `firestore.indexes.json`) that support the composite indexes tested.

### [REQUIRED] Actual behavior

All the tests succeed (even though they shouldn't).

**Comment by nicholaschiang** [comment]
Just FYI the ultimate goal here is to test each possible combination of the different filters (as defined in `FILTERS` and `SORTERS`) available in our app's [search screen](https://tutorbook.app/docs/module-@tutorbook_search-Search.html).

**Comment by samtstern** [comment]
@nicholaschiang thanks for the feedback!  Right now you're correct that the Firestore emulator will execute any query even if the index has not been "created".

@yuchenshi do you know if we will be able to support stricter indexing in the Firestore emulator?

**Comment by yuchenshi** [comment]
Strict indexing in the Firestore Emulator is not a planned feature (yet) but I'll definite discuss with this the team.

On the other hand, we do plan to support dumping all indexes automatically created (say during the test), which you can compare against `firestore.indexes.json`. I believe this will address OP's use case as well.

**Comment by nicholaschiang** [comment]
@yuchenshi That would be a useful (although not **exactly** what I'm asking for) feature; those "tests" would ensure that my `firestore.indexes.json` supported the "tested" queries. Just make sure that feature accounts for [Firestore's ability to merge indexes](https://firebase.google.com/docs/firestore/query-data/index-overview#taking_advantage_of_index_merging) and doesn't just create a composite index for each query type.

**Comment by nicholaschiang** [comment]
Oh, and @samtstern you should have this specification (that the Firestore emulator will execute **any query** even if there isn't an index to support it) added to the [differences between emulator and production webpage](https://firebase.google.com/docs/firestore/security/test-rules-emulator#differences_between_the_emulator_and_production).

**Comment by manwithsteelnerves** [comment]
Its unfortunate that firestore emulator doesn't support index rules.
@yuchenshi I see it useful feature considering the merge will be handled. Thats like a master piece!

What are the chances to consider this feature?

**Comment by yuchenshi** [comment]
We're considering both features of dumping and tracking index rules and we'd appreciate input on the use cases.

As a developer, how do you plan on using emulators to prototype, test, and gain more confidence about your index rules? Shoot us some workflows just like what @nicholaschiang  did and we'll figure out the best way to make that happen.

**Comment by manwithsteelnerves** [comment]
Great!
Our use-case is pretty similar to the OP mentioned.
We have a REST Api's which has optional query parameters. To check if each request has required indexes already in place, we use jest framework to fire the requests with all possible combinations. It throws errors that indexes are missing and we create the required indexes.

So for 4 optional query params with multiple values allowed in each param, we need to test around more than 200 possibilities. And creating indexes for multiple combinations manually is error prone.

If there is a provision for firestore emulator to create the required indexes list, we can just run the all possible cases generated through code and use the indexes created from firestore emulator. This is in-fact save lot of time.

**Comment by voxelbusters** [comment]
Any thing planned related to this feature?

**Comment by manwithsteelnerves** [comment]
@yuchenshi Can you please let us know if anything planned related to this feature?

**Comment by sk-** [comment]
@yuchenshi we just got bitten by this.

We deployed a quick fix, which actually broke later in production due to the missing index. Note that that part of the code was tested, so it'd be great if the emulator had the same behavior as production.

**Comment by yuchenshi** [comment]
@manwithsteelnerves While we don't share timelines and cannot commit to any plans, but I do want to let you know that we hear you and real pain points (including the latest one shared by @sk- ) helps a lot driving our internal discussion. It's a technically challenging and involved feature for sure to get the index semantics match production, but that doesn't mean we're not considering / doing it.

**Comment by manwithsteelnerves** [comment]
As I see it may take more time, I ended up creating 139 manual indexes for my project for all possible test cases. Hopefully, there will be a better solution through emulator workflow!

**Comment by jpangburn** [comment]
Out of the two options discussed "match production" and "dumping all indexes automatically created [during some set of tests]" , if we could only have one then I prefer the second one.  For me, the point of writing a bunch of automated tests is that so later when I add some feature or fix some bug, I don't break a bunch of other things.  With that in mind, if the emulator could dump the list of indexes that I need that aren't already in my firestore.indexes.json- that seems pretty helpful.

If I ran a test set and nothing came out then I'd know that some tests I added (to verify some bug) weren't increasing the indexes I need.  Or if I prototype some feature, it would be nice if the emulator made it clear "hey buddy, you're going to need another 100 indexes to support this" by dumping a list of those hundred indexes and I can realize I had a stupid idea before getting too far into it.

Matching production by giving me an error is kinda useful, but just giving the list of additional indexes needed is going to help me make progress faster I think.  And @nicholaschiang 's comment is critical "Just make sure that feature accounts for Firestore's ability to merge indexes and doesn't just create a composite index for each query type."  Might even be nice if it also said "hey, you've got this set of indexes [X,Y,Z] in your firestore.indexes.json that were not necessary during this test run".

**Comment by merlinnot** [comment]
I'd be in favor of the former (match production) because the latter works only if you have a single application talking to the database - what if you have dozens services in separate repositories talking to the same database?

**Comment by jpangburn** [comment]
> the latter works only if you have a single application talking to the database

Wouldn't you just reset the emulator, run whatever tests you have on your dozens of services in separate repositories, and then end up with the final set of indexes that you need?  I guess I don't see how it's any different than a single application.  Just as a single application is using one Firebase project, presumably your dozens are services are as well?

I guess in the case were you had multiple firebase projects being served by one emulator, you'd want your indexes grouped by project.  But otherwise I don't see a difference- though I could easily be missing something :-)

**Comment by merlinnot** [comment]
I was referring to

> if the emulator could dump the list of indexes that I need that aren't already in my firestore.indexes.json

I assumed you'd like to run your tests and then deploy the indexes in the CI. That wouldn't work in my case, where I have dozens of separate repositories (services) and indexes are stored in a separate repository, shared by all of the services.

So, "if we could only have one", for my use case I'd prefer to load my manually crafted indexes (where you also have optimizations for de-duplication of queries) in all of the services and run test against it. If something fails, I'd add these indexes to my repository with indexes.

Hope that explains my reasoning better 🙂

**Comment by manwithsteelnerves** [comment]
> I assumed you'd like to run your tests and then deploy the indexes in the CI. That wouldn't work in my case, where I have dozens of separate repositories (services) and indexes are stored in a separate repository, shared by all of the services.

To understand better, If you have single firestore db used by multiple services, isn't it possible to run all the diff service tests and update the dumped index file from emulator?

**Comment by merlinnot** [comment]
Not really, no. We have dozens of services interacting with the same database. Each service is developed in a separate repository, by independent teams, deployed individually. We also follow a true to the bone continuous delivery process, so we have multiple (tens/hundreds) of production deployments _daily_.

With that in mind, what you're asking is if for these hundreds deployments daily we can run tests of all our services, instead of just one service, for which the change is made. Let's explore it with an example:

Let's assume 50 services, 300 deployments daily (total across all services), 10 minutes of CI time to run tests for each service. That consumes 300*10=3000 build minutes daily. If we'd run that for all services each time, it's 50 times more, so the cost increase is very significant.

On top of that, what's even worse, some services already run tests on 32-CPU machines (fortunately the emulator supports multithreading). We obviously can't afford a test run that takes 50*10 minutes, so we'd have to run it in parallel. But then we'd need a VM that has 50*32=1600 CPU cores, and such a thing doesn't exist. We could of course run Firestore emulator on a separate VM and execute all tests against it, but the emulator is not very well optimized and it consumes very significant resources for tests of each service, e.g. on a 32 core machine we usually dedicate ~16 cores to the emulator.

For your second point, we could theoretically do that, but that sounds a little complex. I imagine that for each CI run we'd have to pull current Firestore indexes from a central repository, run tests for a service and then validate that all indexes exist. But then we'd be implementing ourselves what we defined here as the first option ("match production"): for the emulator itself to fail when indexes don't exist. And as it was previously mentioned in this very thread, it's not as easy as it might look like, specifically due to features such as [index merging](https://cloud.google.com/firestore/docs/concepts/index-overview#taking_advantage_of_index_merging).

So for me it drills down to implementing "match production" in Firestore or implementing it myself, so I'd obviously prefer it to be implemented in the Emulator.

I also understand your point, I see usefulness of it for small applications, but I believe that with the growth of any application "match production" will eventually become a preference for everyone.

**Comment by manwithsteelnerves** [comment]
Ok, thats really big scale. Its great to see you are able to manage those many projects where I'm struggling to fit firestore in one because of its inconsistent latencies.

So, will dumping the missing indexes compared to production will definitely solve all problems right? It's great to see an index merged dump though. And these indexes will be updated to production manually.

**Comment by jpangburn** [comment]
TL;DR - I agree with the argument for "match production", it would help very large projects a lot.  Would still be nice if there was also a way to generate an optimized list of indexes after running a suite of tests, both for new projects and for large long running projects that may have built up a list of non-optimized indexes.

Loooong answer:
Yes, you clearly have a well oiled machine going there.  The only reason I'm commenting is since they asked us for use cases, I'm trying to help hash it out so they'll have a clear picture (one way or the other) to implement this thing we all need :-)

I was coming at this from the view of having a project with a firestore.indexes.json file that you update with whatever indexes you need, and do a `firestore deploy` and it uploads your indexes, your rules, your functions, etc.  But you aren't doing that @merlinnot , are you? Your devs are writing their own projects, making changes, the CI is running tests for that project, and if the tests pass then it deploys that project's artifacts to production Firestore.  But the CI is not updating any rules or indexes, is it?  This is where the pain is now- the tests can pass but then in production there's a missing index so there's a failure so then someone goes into the Firestore console (or you have your own client that assists you with your flow) and adds that index.  This is why you said:

> I imagine that for each CI run we'd have to pull current Firestore indexes from a central repository, run tests for a service and then validate that all indexes exist

So I think it helps your use case to be clear that when you say "match production", you do NOT mean that some central repository's firestore.indexes.json and firestore.rules is loaded to the emulator.  You mean that the emulator downloads the indexes and rules from production, and proceeds to "match production" as fully as possible.  You need it to behave this way because due to index merging, it's possible that you'll waste indexes by only loading a firestore.indexes.json for a single project (unless they all have completely separated collections and there's no collectionGroups across them).  Is that correct?

That would be pretty useful in your large case because then the CI would fail, and it would fail specifically on the test that was lacking the hypothetical "new needed index".  That would go right back to the developer that wrote it without needing to test the entire set of projects.

For new developing projects, it's nice to have the other way where you just get a list of needed indexes after running all your tests, update your firestore.indexes.json file, and then feel comfortable to deploy to production.  Especially if you could get an optimized list that takes advantage of merging.  But as @merlinnot points out, with the benefit of wisdom from hard knocks I'm sure, is that as projects grow they will begin to prefer "match production".  I can't disagree with that.  The pluses are large in the true CI case with many projects- you get failure right where it belongs.  These large projects are likely where Firebase should optimize (at least for financial benefit).  The downside is you can end up with a very non-optimized set of indexes because you added each as they were needed.

Hope this helps Firebase devs!  Thanks @merlinnot for your very well thought out and detailed description of your use case.

**Comment by samtstern** [comment]
@jpangburn @merlinnot I just want to say this conversation is great and we're listening to both of you!

**Comment by merlinnot** [comment]
@jpangburn By "match production" I mean loading indexes manually myself, not downloading it automatically. I'd prefer that as we strive to keep our development environments fully offline, both for performance and ability to develop under any conditions - aircrafts or simply when one's internet breaks down. We'd probably end up fetching it from a central place and caching, but definitely loading ourselves to the emulator, the same as it's currently done with Firestore Rules. That also allows people to test changes to indexes, we'd definitely have tests in the "Firestore indexes" repository.

At the same time I fully agree that auto-generating indexes would be very helpful for small/medium size applications, would speed up development a lot.

So I guess we can agree that both features would be very useful. Maybe they could be somehow combined, e.g. when some indexes are missing, print helpful messages:

- "A new index is required for this query, add <some JSON> to your indexes.json file."
- "Extending an existing index is required for this query, add <some field> to your <some JSON> index."

We could also have some flag that would result in generating these indexes automatically instead of failing queries.

**Comment by seanburlington** [comment]
I just caught this issue in staging

It's great that the issues list here is so open with clear responses from the dev team - that really saves me time because I now know exactly what this issue is 

It would be great if it was documented here https://firebase.google.com/docs/emulator-suite/connect_firestore

**Comment by samtstern** [comment]
@seanburlington that's a good idea I'll get it documented

**Comment by chrisjacobs91** [comment]
Chipping in on this. Am I right in thinking there is a delay in the emulator building composite indexes? (perhaps as they have not been built in advance like in live environments)

When I change data and then perform a query that would usually require a composite index, I get out of date data back from Firestore.

**Comment by samtstern** [comment]
@chrisjacobs91 that should not be the case and if you can reproduce that please file a separate issue.  The emulator, just like real Firestore, should always be strongly consistent.

**Comment by manwithsteelnerves** [comment]
@samtstern Any updates on this feature? Is it scheduled or not anytime soon?

**Comment by samtstern** [comment]
@manwithsteelnerves I don't have any update here, sorry!

**Comment by Corchoneitor** [comment]
Stumbled with this issue after recently deploying to production (after several previous deploys of the same project). But in this case, I added a "combined query" (which needs a complex indexing). My project did not test indexes, so we stumbled with the error by chance... in production :upside_down_face:.
It'll be helpfull if you could show a warning in the emulator. Or perhaps a config to set emulator with strict/non-strict mode.

**Comment by ddadon10** [comment]
The second solution (dumping all indexes automatically created) is very similar to the [Datastore emulator behavior](https://cloud.google.com/datastore/docs/tools/indexconfig#Datastore_About_index_yaml) and would be a huge benefit for developer productivity.

I also think this information about the emulator behavior deserve to be in the documentation somewhere 🙂

**Comment by IchordeDionysos** [comment]
@ddadon10 it actually is in there: https://firebase.google.com/docs/emulator-suite/connect_firestore#limitations

**Comment by IchordeDionysos** [comment]
@yuchenshi @samtstern we are currently trying to establish a process for data(base) planning and try to establish a process for how we make database decisions, adjust security rules, and indexes.

One way we think about this is that we should write tests for our security rules with database requests our apps make.
So, we always make sure that all Firestore requests in the app will succeed with any changes to the rules.
The same applies to the Firestore indexes where we would also like to test if the requests would be successful.

Right now this wouldn't be possible with the emulator as it does not support emulating the indexes...

The only option I would see is to create a real Firebase project, deploy Firestore with indexes and run the tests against the real database.
However, there are a few issues I see with this approach:
1. Creating indexes is asynchronous
2. Tests for rules and indexes have to behave differently/run as two test runs.
  For example, to enable mocking we would have to deploy an "allow everything"-rule.

**Comment by samtstern** [comment]
@IchordeDionysos fully agree! This is really hard to test and we want to improve it.

**Comment by ziad-saab** [comment]
@samtstern it looks like an internal bug was filed for this on July 2nd. Are there any progress updates that can be shared here?

In appreciation of all the hard work on this project, thank you!

**Comment by samtstern** [comment]
@ziad-saab sorry I'm no longer on the Firebase team ( 😢 🔥 ) so I don't know!

**Comment by alexstanbury** [comment]
I've just run into this problem on production as well. On the web, it's quite easy to handle because you get a nice link in the console which you can click to set up the index for you, but this is for a React Native app on someone's device that doesn't have that luxury. Is there somewhere to find this link within my firebase project?

**Comment by ddadon10** [comment]
@alexstanbury you can try to enable `DATA_READ` on [Firestore audit logging](https://cloud.google.com/firestore/docs/audit-logging) and see if you can get the detailed error message on Cloud Log.

**Comment by alexstanbury** [comment]
> @alexstanbury you can try to enable `DATA_READ` on [Firestore audit logging](https://cloud.google.com/firestore/docs/audit-logging) and see if you can get the detailed error message on Cloud Log.

I appreciate the suggestion but AFAIK that will end up being very expensive as that functionality is billable. I'm hoping there is a simpler/cheaper alternative.

**Comment by dahu33** [comment]
Any update on this? It's unbelievable that in 2022 we can still ship broken code to production because of such issue...

**Comment by lucasilverentand** [comment]
Hi is this feature still being worked on? I very much would like to implement some CI/CD scripts that can validate that the indexes are created using the indexes JSON file.

**Comment by MorenoMdz** [comment]
> Any update on this? It's unbelievable that in 2022 we can still ship broken code to production because of such issue...

Bumping this, the has bitten us a couple of weeks ago.

**Comment by eliezedeck** [comment]
This is a very important feature for the Firestore Emulator. Another one got bit by this, been developing on the emulator for a long while, and I thought everything I did wasn't requiring a specific index until sending a test build to some folks, fortunately, I caught the issue before going into production.

**Comment by go-sean-go** [comment]
Very simple use case: I have a production Firebase environment on Google's cloud, and I don't want to hit it with my dev clients (high risk of screwing up prod data/schemas). Instead, I want a dev env that helps me test & makes it difficult to ship bugs to prod. Not worrying about unit tests, etc - just really basic, normal, modern dev stuff.

Two options:
1. Create a new Firebase app (e.g. "myapp-dev"), or
2. Use local Firebase emulator suite
3. ???

Creating a new dev env (#1) has its own risks - mostly because many Firebase settings are not config-driven (and thus not under git-control, i.e. likely to have small, unintended differences between envs, etc), and #2 has this issue, which (as many have noted above) creates a HIGH likelihood that devs will eventually ship broken queries to prod (and prod seems to fail gracefully, i.e. difficult to troubleshoot). Both can be outage-level stuff. I got hit by #2 in prod today.

Seems to me #1 is bad for the Firebase team (costs Google money?), and bad for devs (aforementioned lack of config file support for many Firebase features). Maybe I'm wrong. Maybe a second app/dev env is the Firebase team recommendation? Would be great to know if that were the case.

Emulator (#2) is overall actually REALLY awesome for me (thanks, Firebase team!), but today's outage (missing index) leaves me real stung. Feels really simple to have some kind of emulator configuration option that would block unindexed queries - this would force/encourage devs to conform their `firestore.indexes.json` to their local app, and deploy it when appropriate, etc. Heck, even just throwing un-blocking console warnings when indexes don't exist would be fantastic. Sort of "proceed at your own risk" stuff. (That also feels like very light lifting to me for the Firebase team, but I know it's never that simple.)

Anyway - hope that's helpful. I wanted to iterate out this VERY basic use case to ensure it's clear.

**Comment by dfink91** [comment]
Run into the issue a couple times as well.

@christhompsongoogle can your self-assignment spark some hope on this issue?

**Comment by christhompsongoogle** [comment]
Nothing to share right now but it's something that we're aware of and are looking at improving in the future.

**Comment by grigb** [comment]
This one issue has cost us many weeks. We now have to completely abandon working locally for a remote dev environment because of this. All future dev will now be done remotely at a huge cost to us!

FIX THIS

**Comment by zanona** [comment]
Not saying this shouldn't be implemented, but, relying purely on a production deployment, rather than having a separate staging project seems a quite risky move IMHO. 🤷‍♂️

**Comment by go-sean-go** [comment]
One other point: Is it not still true that many Firebase settings are ONLY accessible via web console? If I'm wrong, please let me know; I know there have been some advances in the config file lately.

For me, the last time I reviewed the Firebase ecosystem, it was very unfriendly to a Staging environment specifically because it requires a separate project for each environment, and MANY settings are not configurable by config file (i.e. requires a web console to manually configure settings, and cannot easily see the diff between environments). (Maybe this has changed; if not, even one person working alone is bound to make mistakes when trying to keep the two environments in-sync--let alone a whole team with a complex release scheme.)

That is why the Emulator/local dev has an outsized importance in Firebase dev, in my experience--and what makes this issue so severe.

**Editorializing a bit:** I imagine most Firebase users are on small and/or fast-moving teams that don't need or want to worry about Nth-level details in the environment. Syncing anonymized data between environments is not easy, nor is it cheap.

There is of course lots of gray area between a team willing to take some risks vs. a structured team with multiple environments and processes, but I can only imagine the majority of Emulator users are operating without a structured Staging environment (this is 100% speculation).

Worth noting that Firebase does provide this document (which I hadn't seen before today; not sure how old it is): https://firebase.google.com/docs/projects/dev-workflows/overview-environments

For me it is unreasonably difficult in today's Firebase's setup for the reasons mentioned above, but I do hope achieving that as a "one-click solution" is their North star. Depends on their target market, I guess--but, if I cared to implement all that stuff myself, write my own scripts to manage it, etc...well, I probably wouldn't be using Firebase, right? I'd be dangerously close to just rolling my own env to contain costs.

Again - just adding this context here as I hope it helps the Firebase team evaluate the severity of this problem. For me it feels like a broken promise without at least a log warning, or similar. Hope that helps!

**Comment by mcfarljw** [comment]
I'm going to be the first to chime in for getting this supported in 2023. I'm think I'm running in to the same issue as I need a compound query for two properties `deletedAt` and `uploadedAt` for an endpoint the incrementally fetches data that may or may not be deleted.

```ts
const ref = firestore
  .collection('users')
  .doc(user.id)
  .collection('items')
  .where('uploadedAt', '>=', query.offset)

if (query.excludeDeleted)
  ref.where('deletedAt', '==', 0)
```

But I can't test this is working when developing because the emulator completely ignores `ref.where('deletedAt', '==', 0)` when declared with no error or warning and just returns everything after the `uploadedAt` offset.

**Comment by dfink91** [comment]
> I'm going to be the first to chime in for getting this supported in 2023. I'm think I'm running in to the same issue as I need a compound query for two properties `deletedAt` and `uploadedAt` for an endpoint that incrementally fetches data that may or may not be deleted.
> 
> ```ts
> const ref = firestore
>   .collection('users')
>   .doc(user.id)
>   .collection('items')
>   .where('uploadedAt', '>=', query.offset)
> 
> if (query.excludeDeleted)
>   ref.where('deletedAt', '==', 0)
> ```
> 
> But I can't test this is working when developing because the emulator completely ignores `ref.where('deletedAt', '==', 0)` when declared with no error or warning and just returns everything after the `uploadedAt` offset.

as far as I remember, you need to store the result of ref.where(....) somewhere as .where returns a new immutable query. Use that new reference to get the records.

That being said, your problem is not the same as the one discussed here. hope this helps you.

**Comment by mcfarljw** [comment]
> as far as I remember, you need to store the result of ref.where(....) somewhere as .where returns a new immutable query. Use that new reference to get the records.

You are right, it looks like it returns the correct results when using the emulator if I chain them together. I always get this syntax mixed up. Here is my updated example that does query `deletedAt` correctly:

```ts
const ref = firestore
  .collection('users')
  .doc(user.id)
  .collection('items')
  .where('uploadedAt', '>=', query.offset)
  .where('deletedAt', '==', 0)
```

I'll still chime in with my support because even with the change it doesn't recommend indexes and is disconnected from production which makes using it to run tests limited in reliability in terms of catching serious index related issues which is my real pain point at the moment.

**Comment by risalfajar** [comment]
> On the other hand, we do plan to support dumping all indexes automatically created (say during the test), which you can compare against firestore.indexes.json. I believe this will address OP's use case as well.

Tomorrow will be exactly 3 years after this statement. Is there still no progress? Anything? @yuchenshi @taeold 

Currently `firestore.indexes.json` is just useless.

**Comment by garrett-wombat** [comment]
Like @yugisu-flux said. I think its enough to add a warning in the emulator that you've performed a query requiring an index that doesn't exist. Just so the developer will know wheres going to be an issue in production.

**Comment by christhompsongoogle** [comment]
Hi all, I do understand the sentiment here and we haven't forgotten about this item.

If it's helpful in the short term, after performing a compound query you can get the index dump from the firestore emulator here:
http://127.0.0.1:8080/emulator/v1/projects/{projectId}:indexUsage?database=projects/{projectId}/databases/(default)

Example:
```
http://127.0.0.1:8080/emulator/v1/projects/demo-something:indexUsage?database=projects/demo-something/databases/(default)
```
```
{
  "reports": [{
    "index": {
      "name": "projects/demo-something/databases/(default)/collectionGroups/cities/indexes/_",
      "queryScope": "COLLECTION",
      "fields": [{
        "fieldPath": "state",
        "order": "ASCENDING"
      }, {
        "fieldPath": "population",
        "order": "ASCENDING"
      }, {
        "fieldPath": "__name__",
        "order": "ASCENDING"
      }]
    },
    "numQueries": "1"
  }]
}
```

**Disclaimer:** this is unlaunched and the functionality might change in backward-incompatible ways.

**Comment by JZubero** [comment]
> **Disclaimer:** this is unlaunched and the functionality might change in backward-incompatible ways.

Not sure if I am getting it right. Is this supposed to work already or are you proposing this for an upcoming emulator update?

After trying with several emulator setups (where I know that firestore indexes exist) it keeps returning:

```
{
}
```

**Comment by dominicbartl** [comment]
@JZubero Did you actually run queries against the emulator which require an index? The emulator doesn't read your firestore.indexes.json it records the indexes based on your queries.

Also, make sure that you correctly replaced the project ID in the URL. It's in there twice. I got an `{}` response when requesting for a non-existing project.

The endpoint works for me with `firebase-tools@11.30.0`:

```json
{
  "reports": [{
    "index": {
      "name": "projects/myproject-local/databases/(default)/collectionGroups/users/indexes/_",
      "queryScope": "COLLECTION",
      "fields": [{
        "fieldPath": "asd",
        "order": "ASCENDING"
      }, {
        "fieldPath": "test",
        "order": "ASCENDING"
      }, {
        "fieldPath": "__name__",
        "order": "ASCENDING"
      }]
    },
    "numQueries": "3"
  }]
}
```

**Comment by JZubero** [comment]
> Did you actually run queries against the emulator which require an index?

Did the indexes get recorded at runtime only?

I made sure that projectId is valid.

**Comment by dominicbartl** [comment]
> Did the indexes get recorded at runtime only?

Not quite sure what you mean, but you have to execute queries using the emulator for it to add it to its report. If you don't query anything, the emulator won't show any results and return `{}`. Start the emulator and your application, make sure you run the queries and then call the URL

**Comment by konti-kun** [comment]
Thanks for the great feature: index dump!

I tried it, but it seems that this index dump is still logged until it is restarted.
Is there any way to clear it like a REST API?

**Comment by ekalosak** [comment]
Is there a sanctioned place to vote on these features? Getting the unpublished auto-indexing emulator feature to a place where it can update the local firebase.index.json on the fly would be Really Cool, for what it's worth.

**Comment by mono0926** [comment]
@ekalosak 

> Is there a sanctioned place to vote on these features?

https://firebase.uservoice.com/forums/948424-general/suggestions/46593439-firestore-indexes-in-emulator

**Comment by risalfajar** [comment]
> https://firebase.uservoice.com/forums/948424-general/suggestions/46593439-firestore-indexes-in-emulator

This is the first time I heard of this site, from where did you know it?

**Comment by mono0926** [comment]
@risalfajar 


> > https://firebase.uservoice.com/forums/948424-general/suggestions/46593439-firestore-indexes-in-emulator
> 
> This is the first time I heard of this site, from where did you know it?


Here: https://firebase.google.com/community

<img width="404" alt="image" src="https://github.com/firebase/firebase-tools/assets/1255062/35c689bb-3419-4cfe-a172-1eb1250f23c3">

Previously, it was private information within the [Firebase Alpha Program](https://services.google.com/fb/forms/firebasealphaprogram/) Google Groups.

**Comment by lukepighetti** [comment]
This is absolutely critical. We currently have no way to warn developers that they are creating functionality that requires indexes in production. All we expect is a warning in the console just like we experience when developing on a managed Firestore instance.

**Comment by jaredweinfurtner** [comment]
it is now 2024 and still no solution.  @christhompsongoogle, do you have any update?

**Comment by omervexler-fijoya** [comment]
@christhompsongoogle I really ❤️ Google and firebase, would love to receive updates if there are any. Thanks!

**Comment by christhompsongoogle** [comment]
Hi all, unfortunately I'm unable to provide any timeline for this feature. If this is critical to your application, the only way to know for sure that a query is covered is testing the indexes+queries against a production instance of Firestore, in a test project for example.

**Comment by aetherknight** [comment]
This feature (or rather, this feature but for datastore-mode) is necessary for us to be able to move off of the older datastore-emulator. Switching to the newer firestore emulator is also necessary for us to be able to move to the latest version of NDB (2.3.1).

We rely heavily on unit testing to ensure that all queries work with our indexes -- it is not practical to be able to test every possible query combination and every code path in a deployed environment after every code change. Our code base is large and complex, and many queries are fairly deep in our code base and cannot be directly or easily called. They often have edge cases that are easy to test with unit tests, but are not easily tested through integration testing in a dev or staging environment.

**Comment by csilvers** [comment]
@christhompsongoogle thank you for the pointer to the `indexUsage` REST request, that would be sufficient for our needs.  Sadly, it does not seem to work in datastore-mode, only in firestore-mode:
```
% curl -D- localhost:8889/emulator/v1/projects/'khan-test:indexUsage?database=projects/khan-test/databases/(default)'
HTTP/1.1 400 Bad Request
content-type: application/json
content-length: 169

{"error":{"code":400,"message":"Database mode not set to Firestore Mode. If using the gCloud CLI, please specify the --database-mode flag.","status":"INVALID_ARGUMENT"}}
```

Is there any plan -- and ideally timeline -- to support this functionality in datastore mode?

**Comment by ericvera** [comment]
@christhompsongoogle Thanks for the info on the emulator report! 

Is it expected to also log missing indexes from the client SDK? or only from the admin SDK?

I built a small CLI to simplify/automate updating/printing out the index file ([firestore-index-gen](https://www.npmjs.com/package/firestore-index-gen)), but I am noticing no logs of indexes for queries that go directly to Firestore from the client SDK. I am curious whether this is by design or a limitation.

**Comment by downright-development** [comment]
+1 - Just hit this myself, had no clue I was missing indexes until my deployment hit prod.
Admit - shame on me for not knowing this first hand but feels like this defeats the purpose of the emulator.

**Comment by digimbyte** [comment]
> [@christhompsongoogle](https://github.com/christhompsongoogle) Thanks for the info on the emulator report!
> 
> In case it is useful to someone else, I built a small CLI to simplify/automate updating/printing out the index file ([firestore-index-gen](https://www.npmjs.com/package/firestore-index-gen)).

this is probablky the best solution. separates dev env from index captures. allowing finer control before going into prod

**Comment by jorgealemangonzalez** [comment]
@ericvera Build an amazing tool, but as far as I saw it depends on tests using all indexes, as otherwise override will delete the indexes not used by the tests. 

Did anyone solved this problem?

**Comment by eliezedeck** [comment]
The way I see this, is that you'll need to create another Firestore database. You'll run your pre-prod tests there.

That'll force you to have a billing account, which, at the end of the day, is what Google wants. There are probably too many free Firestore databases out there that's not monetized enough, so they have to disregard all efforts that would slow down conversion.

It's always about the economics.

**Comment by risalfajar** [comment]
First comment in 2026, still no news.

**Comment by kdawgwilk** [comment]
Well I guess there is some news https://firebase.blog/posts/2026/01/firestore-enterprise-pipeline-operations/ Apparently firestore enterprise edition does not require indexes to query data anymore so your queries won't be blocked by missing indexes?  At least it seems that way if i am reading this right

>Firestore Enterprise edition doesn’t index collections by default, nor does it require a collection be indexed in order for it to be queried.
